File: backend/app/services/monitor_service.py
from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator
from datetime import UTC, datetime, timedelta

import psutil
from sqlalchemy.ext.asyncio import AsyncSession

from app.repositories.monitor_repo import MonitorRepo


class MonitorService:
    def __init__(
        self,
        repo: MonitorRepo,
    ) -> None:
        self.repo = repo

    @staticmethod
    def _get_server_status_payload() -> dict[str, float | int | None]:
        cpu_percent = psutil.cpu_percent(interval=1)
        cpu_cores = psutil.cpu_count(logical=True)

        mem = psutil.virtual_memory()
        mem_total = round(mem.total / 1024 / 1024)
        mem_used = round(mem.used / 1024 / 1024)
        mem_usage = round(mem.percent, 2)

        disk = psutil.disk_usage("/")
        disk_total = round(disk.total / 1024 / 1024 / 1024, 2)
        disk_used = round(disk.used / 1024 / 1024 / 1024, 2)
        disk_usage = round(disk.percent, 2)

        return {
            "cpu_percent": cpu_percent,
            "cpu_cores": cpu_cores,
            "mem_total": mem_total,
            "mem_used": mem_used,
            "mem_usage": mem_usage,
            "disk_total": disk_total,
            "disk_used": disk_used,
            "disk_usage": disk_usage,
        }

    async def get_overview(
        self, session: AsyncSession, days: int
    ) -> dict:
        end_time = datetime.now(UTC)
        start_time = end_time - timedelta(days=days)

        repo = self.repo

        total_visits = await repo.count_visits_since(session, start_time)
        unique_visitors = await repo.count_unique_visitors_since(session, start_time)
        unique_visitor_ids = await repo.count_unique_visitor_ids_since(
            session, start_time
        )
        top_pages = await repo.get_top_pages_since(session, start_time, limit=10)
        browser_stats = await repo.get_browser_stats_since(session, start_time)
        os_stats = await repo.get_os_stats_since(session, start_time)
        daily_trend = await repo.get_daily_trend_since(session, start_time)

        return {
            "total_visits": total_visits,
            "unique_visitors": unique_visitors,
            "unique_visitor_ids": unique_visitor_ids,
            "top_pages": top_pages,
            "browser_stats": browser_stats,
            "os_stats": os_stats,
            "daily_trend": daily_trend,
            "period_days": days,
        }

    async def get_visitors(
        self, session: AsyncSession, days: int, page: int, page_size: int
    ) -> dict:
        end_time = datetime.now(UTC)
        start_time = end_time - timedelta(days=days)

        repo = self.repo

        total = await repo.count_visits_since(session, start_time)
        offset = (page - 1) * page_size
        visitors = await repo.list_visitors_since(
            session,
            start_time,
            offset=offset,
            limit=page_size,
        )

        visitor_list = [
            {
                "id": v.id,
                "visitor_id": v.visitor_id,
                "page_url": v.page_url,
                "page_path": v.page_path,
                "referrer": v.referrer,
                "browser": v.browser,
                "screen_resolution": v.screen_resolution,
                "language": v.language,
                "ip_address": v.ip_address,
                "visit_time": v.visit_time.isoformat()
                if v.visit_time
                else None,
            }
            for v in visitors
        ]

        return {
            "list": visitor_list,
            "total": total,
            "page": page,
            "page_size": page_size,
            "total_pages": (total + page_size - 1) // page_size,
        }

    async def get_user_logins(
        self,
        session: AsyncSession,
        days: int,
        page: int,
        page_size: int,
    ) -> dict:
        end_time = datetime.now(UTC)
        start_time = end_time - timedelta(days=days)

        repo = self.repo

        users = await repo.list_users_with_login_records(session)

        login_logs = []
        for user in users:
            has_recent_login = False
            if user.current_login_at and user.current_login_at >= start_time:
                has_recent_login = True
            if user.last_login_at and user.last_login_at >= start_time:
                has_recent_login = True

            if has_recent_login:
                login_logs.append(
                    {
                        "user_id": user.id,
                        "username": user.username,
                        "name": user.name,
                        "login_count": user.login_count,
                        "last_login_at": (
                            user.last_login_at.isoformat()
                            if user.last_login_at
                            else None
                        ),
                        "current_login_at": (
                            user.current_login_at.isoformat()
                            if user.current_login_at
                            else None
                        ),
                        "last_login_ip": user.last_login_ip,
                        "current_login_ip": user.current_login_ip,
                        "active": user.active,
                    }
                )

        total = len(login_logs)
        offset = (page - 1) * page_size
        paginated_logs = login_logs[offset : offset + page_size]

        return {
            "list": paginated_logs,
            "total": total,
            "page": page,
            "page_size": page_size,
            "total_pages": (total + page_size - 1) // page_size,
        }

    async def get_server_status(self) -> dict[str, float | int | None]:
        return self._get_server_status_payload()

    async def stream_server_status(
        self,
    ) -> AsyncIterator[dict[str, float | int | None]]:
        while True:
            yield self._get_server_status_payload()
            await asyncio.sleep(5)

    async def get_daily_summary(
        self, session: AsyncSession, date: datetime
    ) -> dict:
        """获取指定日期的访问统计摘要。

        Args:
            date: 日期（取其当天 00:00 ~ 次日 00:00）

        Returns:
            包含总访问量、独立访客、热门页面等统计信息的字典
        """
        start = date.replace(hour=0, minute=0, second=0, microsecond=0)
        end = start + timedelta(days=1)

        repo = self.repo

        total_visits = await repo.count_visits_between(session, start, end)
        unique_visitors = await repo.count_unique_visitors_between(session, start, end)
        unique_ips = await repo.count_unique_ips_between(session, start, end)
        top_pages = await repo.get_top_pages_between(session, start, end, limit=5)
        browser_stats = await repo.get_browser_stats_between(session, start, end)
        os_stats = await repo.get_os_stats_between(session, start, end)
        device_stats = await repo.get_device_stats_between(session, start, end)

        return {
            "date": start.strftime("%Y-%m-%d"),
            "total_visits": total_visits,
            "unique_visitors": unique_visitors,
            "unique_ips": unique_ips,
            "top_pages": top_pages,
            "browser_stats": browser_stats,
            "os_stats": os_stats,
            "device_stats": device_stats,
        }

File: go-backend/internal/service/monitor_service.go
package service

import (
	"context"
	"math"
	"runtime"
	"time"

	"github.com/shirou/gopsutil/v4/cpu"
	"github.com/shirou/gopsutil/v4/disk"
	"github.com/shirou/gopsutil/v4/load"
	"github.com/shirou/gopsutil/v4/mem"

	"github.com/KanoCifer/kuroome-blog/internal/dto"
	"github.com/KanoCifer/kuroome-blog/internal/model"
	"github.com/KanoCifer/kuroome-blog/internal/repository/postgres"
)

// Monitorer 定义 monitor handler 依赖的能力集合。
// 由 service.Monitorer 实现；handler 仅依赖此接口，便于测试替换。
type Monitorer interface {
	GetOverview(ctx context.Context, days int) (dto.OverviewResponse, error)
	GetVisitors(ctx context.Context, days, page, pageSize int) (dto.VisitorsResponse, error)
	GetUserLogins(ctx context.Context, days, page, pageSize int) (dto.UserLoginsResponse, error)
	GetServerStatus() (dto.ServerStatusResponse, error)
	StreamServerStatus(ctx context.Context) (<-chan dto.ServerStatusResponse, error)
	TrackVisitor(ctx context.Context, data dto.VisitorResponse) error
	GetStatusDetail(ctx context.Context) (dto.StatusDetailResponse, error)
}

// MonitorService 实现 monitor 端点的业务逻辑（overview / visitors / user-logins
// / status-detail）。
type MonitorService struct {
	visitor   *postgres.VisitorRepo
	user      *postgres.UserRepo
	version   string
	startTime time.Time
}

func NewMonitorService(visitor *postgres.VisitorRepo, user *postgres.UserRepo, version string) *MonitorService {
	return &MonitorService{visitor: visitor, user: user, version: version, startTime: time.Now()}
}

// TrackVisitor 记录访客追踪数据（公开接口）。visit_time 由 PG default current_timestamp 填充。
func (s *MonitorService) TrackVisitor(ctx context.Context, data dto.VisitorResponse) error {
	track := &model.VisitorTrack{
		VisitorID:        data.VisitorID,
		PageURL:          data.PageURL,
		PagePath:         data.PagePath,
		Referrer:         ptrIf(data.Referrer),
		Browser:          ptrIf(data.Browser),
		ScreenResolution: ptrIf(data.ScreenResolution),
		Language:         ptrIf(data.Language),
		IPAddress:        data.IpAddress,
		BrowserName:      ptrIf(data.BrowserName),
		BrowserVersion:   ptrIf(data.BrowserVersion),
		OSName:           ptrIf(data.OSName),
		OSVersion:        ptrIf(data.OSVersion),
		CPU:              ptrIf(data.Cpu),
		DeviceType:       ptrIf(data.DeviceType),
	}
	return s.visitor.Insert(ctx, track)
}

// GetStatusDetail 返回服务器运行时状态概览，对齐 Python
// PublicService.get_status_detail 的 {version, service, system} 三元组。
func (s *MonitorService) GetStatusDetail(ctx context.Context) (dto.StatusDetailResponse, error) {
	// --- system info --------------------------------------------------- //
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	cpuCountL, _ := cpu.Counts(true)
	cpuCountP, _ := cpu.Counts(false)

	cpuPercent := 0.0
	if p, err := cpu.Percent(0, false); err == nil && len(p) > 0 {
		cpuPercent = math.Round(p[0])
	}

	loadAvg := &load.AvgStat{}
	if la, err := load.Avg(); err == nil {
		loadAvg = la
	}

	// 2. 获取CPU信息
	cpuInfo, _ := cpu.Info()

	vm, _ := mem.VirtualMemory()

	sysInfo := dto.SystemInfoResponse{
		SystemTime:       time.Now().Format("2006/01/02 15:04:05"),
		SystemTimezone:   "GMT+8",
		OsName:           runtime.GOOS,
		OsVersion:        runtime.GOARCH,
		KernelVersion:    runtime.GOARCH,
		CpuModel:         cpuInfo[0].ModelName,
		CpuCountPhysical: cpuCountP,
		CpuCountLogical:  cpuCountL,
		LoadAverage: map[string]float64{
			"1m":  math.Round(loadAvg.Load1*100) / 100,
			"5m":  math.Round(loadAvg.Load5*100) / 100,
			"15m": math.Round(loadAvg.Load15*100) / 100,
		},
		CpuPercent:         cpuPercent,
		MemoryUsagePercent: round2(vm.UsedPercent),
		MemoryUsedBytes:    vm.Used,
		MemoryTotalBytes:   vm.Total,
	}

	// --- service info -------------------------------------------------- //
	dbOk := true
	if err := s.visitor.Ping(ctx); err != nil {
		dbOk = false
	}

	svcInfo := dto.ServiceInfoResponse{
		Runtime:          runtime.Version() + " " + runtime.GOOS + "/" + runtime.GOARCH,
		GoVersion:        runtime.Version(),
		Goroutines:       runtime.NumGoroutine(),
		GcCount:          m.NumGC,
		StartTime:        s.startTime.Unix(),
		HeapMemoryBytes:  m.Alloc,
		TotalMemoryBytes: m.TotalAlloc,
		DbOk:             dbOk,
		ApiOk:            true,
	}

	// --- version info -------------------------------------------------- //
	verInfo := dto.VersionInfoResponse{
		RepoURL:        "https://github.com/KanoCifer/kuroome-blog",
		CurrentVersion: s.version,
	}

	return dto.StatusDetailResponse{
		Version: verInfo,
		Service: svcInfo,
		System:  sysInfo,
	}, nil
}

// GetOverview 返回 days 天内的访客总览统计，对齐 Python MonitorService.get_overview。
func (s *MonitorService) GetOverview(ctx context.Context, days int) (dto.OverviewResponse, error) {
	endTime := time.Now().UTC()
	startTime := endTime.AddDate(0, 0, -days)

	totalVisits, err := s.visitor.CountVisitsSince(ctx, startTime)
	if err != nil {
		return dto.OverviewResponse{}, err
	}
	uniqueVisitors, err := s.visitor.CountUniqueVisitorsSince(ctx, startTime)
	if err != nil {
		return dto.OverviewResponse{}, err
	}
	uniqueVisitorIDs, err := s.visitor.CountUniqueVisitorIDsSince(ctx, startTime)
	if err != nil {
		return dto.OverviewResponse{}, err
	}
	topPages, err := s.visitor.GetTopPagesSince(ctx, startTime, 10)
	if err != nil {
		return dto.OverviewResponse{}, err
	}
	browserStats, err := s.visitor.GetBrowserStatsSince(ctx, startTime)
	if err != nil {
		return dto.OverviewResponse{}, err
	}
	osStats, err := s.visitor.GetOSSStatsSince(ctx, startTime)
	if err != nil {
		return dto.OverviewResponse{}, err
	}
	dailyTrend, err := s.visitor.GetDailyTrendSince(ctx, startTime)
	if err != nil {
		return dto.OverviewResponse{}, err
	}

	return dto.OverviewResponse{
		TotalVisits:      totalVisits,
		UniqueVisitors:   uniqueVisitors,
		UniqueVisitorIDs: uniqueVisitorIDs,
		TopPages:         topPages,
		BrowserStats:     browserStats,
		OSSStats:         osStats,
		DailyTrend:       dailyTrend,
		PeriodDays:       days,
	}, nil
}

// GetVisitors 返回 days 天内的访客分页列表，对齐 Python MonitorService.get_visitors。
func (s *MonitorService) GetVisitors(ctx context.Context, days, page, pageSize int) (dto.VisitorsResponse, error) {
	endTime := time.Now().UTC()
	startTime := endTime.AddDate(0, 0, -days)

	total, err := s.visitor.CountVisitsSince(ctx, startTime)
	if err != nil {
		return dto.VisitorsResponse{}, err
	}

	offset := (page - 1) * pageSize
	tracks, err := s.visitor.ListVisitorsSince(ctx, startTime, offset, pageSize)
	if err != nil {
		return dto.VisitorsResponse{}, err
	}

	list := make([]dto.VisitorItem, 0, len(tracks))
	for _, v := range tracks {
		list = append(list, dto.VisitorItem{
			ID:               v.ID,
			VisitorID:        v.VisitorID,
			PageURL:          v.PageURL,
			PagePath:         v.PagePath,
			Referrer:         v.Referrer,
			Browser:          v.Browser,
			ScreenResolution: v.ScreenResolution,
			Language:         v.Language,
			IPAddress:        v.IPAddress,
			VisitTime:        &v.VisitTime,
		})
	}

	return dto.VisitorsResponse{
		List:       list,
		Total:      total,
		Page:       page,
		PageSize:   pageSize,
		TotalPages: (total + pageSize - 1) / pageSize,
	}, nil
}

// GetUserLogins 返回 days 天内有登录记录的用户分页列表，对齐 Python MonitorService.get_user_logins。
func (s *MonitorService) GetUserLogins(ctx context.Context, days, page, pageSize int) (dto.UserLoginsResponse, error) {
	endTime := time.Now().UTC()
	startTime := endTime.AddDate(0, 0, -days)

	users, err := s.user.ListUsersWithLoginRecords(ctx)
	if err != nil {
		return dto.UserLoginsResponse{}, err
	}

	logins := make([]dto.UserLoginItem, 0, len(users))
	for _, u := range users {
		if !hasRecentLogin(u, startTime) {
			continue
		}
		logins = append(logins, dto.UserLoginItem{
			UserID:         int(u.ID),
			Username:       u.Username,
			Name:           u.Name,
			LoginCount:     u.LoginCount,
			LastLoginAt:    isoPtr(u.LastLoginAt),
			CurrentLoginAt: isoPtr(u.CurrentLoginAt),
			LastLoginIP:    u.LastLoginIP,
			CurrentLoginIP: u.CurrentLoginIP,
			Active:         u.Active,
		})
	}

	total := len(logins)
	offset := (page - 1) * pageSize
	end := offset + pageSize
	if end > total {
		end = total
	}
	if offset > total {
		offset = total
	}
	paginated := logins[offset:end]

	return dto.UserLoginsResponse{
		List:       paginated,
		Total:      total,
		Page:       page,
		PageSize:   pageSize,
		TotalPages: (total + pageSize - 1) / pageSize,
	}, nil
}

// hasRecentLogin 判断用户是否在时间窗内有登录记录（对齐 Python 同级逻辑）。
func hasRecentLogin(u model.User, start time.Time) bool {
	if u.CurrentLoginAt != nil && !u.CurrentLoginAt.Before(start) {
		return true
	}
	if u.LastLoginAt != nil && !u.LastLoginAt.Before(start) {
		return true
	}
	return false
}

// isoPtr 将 time.Time 转为 ISO 字符串指针，nil 则返回 nil（对齐 Python .isoformat() 语义）。
func isoPtr(t *time.Time) *string {
	if t == nil {
		return nil
	}
	s := t.UTC().Format(time.RFC3339)
	return &s
}

// GetServerStatus 返回当前服务器 CPU/内存/磁盘指标，对齐 Python
// _get_server_status_payload 的字段名、单位与舍入规则。
//
// cpu.Percent 用 1s 间隔（与 Python psutil.cpu_percent(interval=1) 一致），
// 会阻塞约 1s 换取更准的 CPU 利用率。
func (s *MonitorService) GetServerStatus() (dto.ServerStatusResponse, error) {
	cpuPercents, err := cpu.Percent(time.Second, false)
	if err != nil {
		return dto.ServerStatusResponse{}, err
	}
	var cpuPercent float64
	if len(cpuPercents) > 0 {
		cpuPercent = cpuPercents[0]
	}

	cpuCores, err := cpu.Counts(true)
	if err != nil {
		return dto.ServerStatusResponse{}, err
	}

	vm, err := mem.VirtualMemory()
	if err != nil {
		return dto.ServerStatusResponse{}, err
	}

	du, err := disk.Usage("/")
	if err != nil {
		return dto.ServerStatusResponse{}, err
	}

	return dto.ServerStatusResponse{
		CPUPercent: math.Round(cpuPercent),
		CPUCores:   cpuCores,
		MemTotal:   int(math.Round(float64(vm.Total) / 1024 / 1024)),
		MemUsed:    int(math.Round(float64(vm.Used) / 1024 / 1024)),
		MemUsage:   round2(vm.UsedPercent),
		DiskTotal:  round2(float64(du.Total) / 1024 / 1024 / 1024),
		DiskUsed:   round2(float64(du.Used) / 1024 / 1024 / 1024),
		DiskUsage:  round2(du.UsedPercent),
	}, nil
}

// StreamServerStatus 每 5s 推送一帧 server status，直到 ctx 取消。
// 对齐 Python stream_server_status：一个无限循环，每 5s yield 一次 payload。
func (s *MonitorService) StreamServerStatus(ctx context.Context) (<-chan dto.ServerStatusResponse, error) {
	ch := make(chan dto.ServerStatusResponse)
	go func() {
		defer close(ch)
		ticker := time.NewTicker(5 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				status, err := s.GetServerStatus()
				if err != nil {
					continue
				}
				select {
				case ch <- status:
				case <-ctx.Done():
					return
				}
			}
		}
	}()
	return ch, nil
}

// round2 保留两位小数（对齐 Python round(x, 2) 语义）。
func round2(x float64) float64 {
	return math.Round(x*100) / 100
}
